modern/org/commcare/modern/database/TableBuilder.java

```diff
@@ -0,0 +1,188 @@
+import java.io.OutputStream;
+import java.lang.reflect.Field;
+import java.lang.reflect.Method;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Vector;
+
+/**
+ * Functions for generating CommCare SQL statements based on classes
+ *
+ * Largely taken from renamed AndroidTableBuilder and moved into api to be used
+ * externally.
+ *
+ * @author ctsims
+ * @author wspride
+ */
+public class TableBuilder {
+
+    private String name;
+
+    private Vector<String> cols;
+    private Vector<String> rawCols;
+
+    public TableBuilder(Class c, String name) {
+        this.name = name;
+        cols = new Vector<String>();
+        rawCols = new Vector<String>();
+        this.addData(c);
+    }
+
+    public TableBuilder(String name) {
+        this.name = name;
+        cols = new Vector<String>();
+        rawCols = new Vector<String>();
+    }
+
+    public void addData(Class c) {
+        cols.add(DatabaseHelper.ID_COL + " INTEGER PRIMARY KEY");
+        rawCols.add(DatabaseHelper.ID_COL);
+
+        for(Field f : c.getDeclaredFields()) {
+            if(f.isAnnotationPresent(MetaField.class)) {
+                MetaField mf = f.getAnnotation(MetaField.class);
+                addMetaField(mf);
+            }
+        }
+
+        for(Method m : c.getDeclaredMethods()) {
+            if(m.isAnnotationPresent(MetaField.class)) {
+                MetaField mf = m.getAnnotation(MetaField.class);
+                addMetaField(mf);
+            }
+        }
+
+        cols.add(DatabaseHelper.DATA_COL + " BLOB");
+        rawCols.add(DatabaseHelper.DATA_COL);
+    }
+
+
+    protected void addMetaField(MetaField mf) {
+        String key = mf.value();
+        String columnName = scrubName(key);
+        rawCols.add(columnName);
+        String columnDef;
+        columnDef = columnName;
+
+        //Modifiers
+        if(unique.contains(columnName) || mf.unique()) {
+            columnDef += " UNIQUE";
+        }
+        cols.add(columnDef);
+    }
+
+    public void addData(Persistable p) {
+        cols.add(DatabaseHelper.ID_COL + " INTEGER PRIMARY KEY");
+        rawCols.add(DatabaseHelper.ID_COL);
+
+        if(p instanceof IMetaData) {
+            String[] keys = ((IMetaData)p).getMetaDataFields();
+            for(String key : keys) {
+                String columnName = scrubName(key);
+                if(!rawCols.contains(columnName)) {
+                    rawCols.add(columnName);
+                    String columnDef = columnName;
+
+                    //Modifiers
+                    if (unique.contains(columnName)) {
+                        columnDef += " UNIQUE";
+                    }
+                    cols.add(columnDef);
+                }
+            }
+        }
+
+        cols.add(DatabaseHelper.DATA_COL + " BLOB");
+        rawCols.add(DatabaseHelper.DATA_COL);
+    }
+
+
+    HashSet<String> unique = new HashSet<String>();
```

Comment:
Can this be made `private`? Also, please put all class fields at the top of the class definition; helps a lot with readability.